src/inmanta/server/services/userservice.py

```diff
@@ -0,0 +1,110 @@
+"""
+    Copyright 2023 Inmanta
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+        http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+
+    Contact: code@inmanta.com
+"""
+import logging
+
+import nacl.exceptions
+import nacl.pwhash
+from inmanta import const, data, protocol
+from inmanta.data import model
+from inmanta.protocol import common, exceptions
+from inmanta.server import SLICE_TRANSPORT, SLICE_USER
+from inmanta.server import protocol as server_protocol
+
+LOGGER = logging.getLogger(__name__)
+
+
+class UserService(server_protocol.ServerSlice):
+    """Slice for managing users"""
+
+    def __init__(self) -> None:
+        super().__init__(SLICE_USER)
+
+    def get_depended_by(self) -> list[str]:
+        return [SLICE_TRANSPORT]
+
+    @protocol.handle(protocol.methods_v2.list_users)
+    async def list_users(self) -> list[model.User]:
+        return [user.to_dao() for user in await data.User.get_list()]
+
+    @protocol.handle(protocol.methods_v2.add_user)
+    async def add_user(self, username: str, password: str) -> model.User:
+        # check if the user already exists
+        existing = await data.User.get_one(username=username)
+        if existing:
+            raise exceptions.Conflict(f"A user with name {username} already exists.")
```

Comment:
There is a race condition here if the new user is created between the exists check and insert. In that case the endpoint will crash with a 500 response code.

Reply:
This should be done in a single transaction right?

Reply:
We could go for a transaction here, but that would mean we need to use the transaction isolation level repeatable reads. Those types of transactions can fails with an exception when the database fails to serialize the actions on the database. That means we have to implement a retry mechanism as well, which feels like overkill to me. Alternatively we can just perform the insert and catch the exception that occurs when there is a primary key conflict.

Reply:
I'll go for the try/catch :p

Reply:
I added the try/except (not pushed yet) put it seems hard to write a test for this